src/v/http/request_builder.h

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+#include "thirdparty/ada/ada.h"
+
+#include <seastar/core/sstring.hh>
+
+#include <absl/container/flat_hash_map.h>
+#include <absl/container/flat_hash_set.h>
+#include <boost/beast/http/message.hpp>
+
+namespace http {
+
+// Builds a request using the builder pattern. Allows setting the host, target
+// (path), method, headers and query params.
+class request_builder {
+public:
+    using error_type = ss::sstring;
+    static constexpr auto default_state{"host not set"};
+
+    using expected
+      = tl::expected<boost::beast::http::request_header<>, error_type>;
```

Comment:
where is `tl::`?

Reply:
This is the `expected` library from https://github.com/TartanLlama/expected, it is a backport of https://en.cppreference.com/w/cpp/utility/expected from C++23 with some extension methods.

In our build it comes from https://github.com/ada-url/ada which vendors the header https://github.com/ada-url/ada/blob/main/include/ada/expected.h

This is the type I was talking about last week in standup when talking about chaining error handling operations.